src/plugins/autofill/autofill.js

```diff
@@ -231,19 +234,49 @@ export class Autofill extends BasePlugin {
 
     if (startOfDragCoords && startOfDragCoords.row > -1 && startOfDragCoords.col > -1) {
       const selectionData = this.getSelectionData();
-      const beforeAutofillHook = this.hot.runHooks('beforeAutofill', startOfDragCoords, endOfDragCoords, selectionData);
 
-      if (beforeAutofillHook === false) {
+      const sourceRange = {
+        from: {
+          row: Math.min(selectionRangeLast.from.row, selectionRangeLast.to.row),
+          col: Math.min(selectionRangeLast.from.col, selectionRangeLast.to.col)
+        },
+        to: {
+          row: Math.max(selectionRangeLast.from.row, selectionRangeLast.to.row),
+          col: Math.max(selectionRangeLast.from.col, selectionRangeLast.to.col)
+        }
+      };
+
+      const targetRange = {
+        from: {
+          row: Math.min(startOfDragCoords.row, endOfDragCoords.row),
+          col: Math.min(startOfDragCoords.col, endOfDragCoords.col)
+        },
+        to: {
+          row: Math.max(startOfDragCoords.row, endOfDragCoords.row),
+          col: Math.max(startOfDragCoords.col, endOfDragCoords.col)
+        }
+      };
+
+      const beforeAutofillHookResult = this.hot.runHooks(
+        'beforeAutofill',
+        selectionData,
+        sourceRange,
+        targetRange,
+        directionOfDrag
+      );
+
+      if (beforeAutofillHookResult === false) {
         this.hot.selection.highlight.getFill().clear();
         this.hot.render();
 
         return false;
       }
 
       const deltas = getDeltas(startOfDragCoords, endOfDragCoords, selectionData, directionOfDrag);
-      let fillData = selectionData;
 
-      if (['up', 'left'].indexOf(directionOfDrag) > -1) {
+      let fillData = beforeAutofillHookResult;
+
+      if (['up', 'left'].indexOf(directionOfDrag) > -1 && beforeAutofillHookResult === selectionData) {
```

Comment:
That may be confusing. Once you will get an array in `beforeAutofill` hook that will be inverted, some other time, it will not. The data may hop through few such hooks and you can't differentiate what will happen with the one you got

Reply:
I meant that as you write, `beforeAutofill` will always have not-inverted data. But after that it was unpredictable. If there was a hook that returned (even the same data, but different array identity), you can't know how to modify the data / if it will be inverted. There may be multiple hooks in the pipeline:

```ts
const slectedData = [ ['a', 'b'], ['c', 'd'] ];

addHook('beforeAutofill', (data) => {
  return [ ['a', 'b'], ['c', 'd'] ];
})

addHook('beforeAutofill', (data) => {
  data[0][0] = 'x';

  // author of this hook does not know that the data will not be inverted
})
```

As you did it now, it will always be inverted. Makes thinking about the outcome easier and predictable IMO.

Reply:
Ah I see, thanks